### PR TITLE
New BRS + Autoremediation for Minimum TLS Storage

### PR DESCRIPTION
**Pull request by abhikash01**

### CODE REVIEW

File: Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1 [hunk 1]
```diff
@@ -0,0 +1,959 @@
...
+            try
+            {
+                $name = $_.ResourceName
+                $resourceGroupName = $_.ResourceGroupName
+            
+                $resStorageAccount = Get-AzStorageAccount  -Name $name  -ResourceGroupName $resourceGroupName -ErrorAction SilentlyContinue
+                $storageAccountResources = $resStorageAccount | Select-Object @{N='StorageAccountName';E={$_.StorageAccountName}},
+                                                                            @{N='ResourceGroupName';E={$_.ResourceGroupName}},
+                                                                            @{N='PrimaryLocation';E={$_.PrimaryLocation}},
+                                                                            @{Name ='MinimalTlsVersion';Expression={ (Get-AzStorageAccount -Name $_.StorageAccountName -ResourceGroupName $_.ResourceGroupName).MinimumTlsVersion }}
```
Comment (aakashpahwa-msft) [inline Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1:299]: Why are we doing Get-AzStorageAccount twice?

File: Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1 [hunk 2]
```diff
@@ -0,0 +1,959 @@
...
+        if ([String]::IsNullOrWhiteSpace($FilePath))
+        {
+            Write-Host "`nFetching all Storage Accounts in Subscription: $($context.Subscription.SubscriptionId)" -ForegroundColor $([Constants]::MessageType.Info)
+            Write-Host $([Constants]::SingleDashLine)
+            # Get all Storage Accounts in the Subscription
+            $storageAccounts = Get-AzStorageAccount  -ErrorAction Stop
+            $storageAccountResources = $storageAccounts | Select-Object @{N='StorageAccountName';E={$_.StorageAccountName}},
+                                                                        @{N='ResourceGroupName';E={$_.ResourceGroupName}},
+                                                                        @{N='PrimaryLocation';E={$_.PrimaryLocation}},
+                                                                        @{Name ='MinimalTlsVersion';Expression={ (Get-AzStorageAccount -Name $_.StorageAccountName -ResourceGroupName $_.ResourceGroupName).MinimumTlsVersion }}
```
Comment (aakashpahwa-msft) [inline Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1:327]: Same question as above.

File: Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1 [hunk 3]
```diff
@@ -0,0 +1,959 @@
...
+            $validstorageAccountResources | ForEach-Object {
+                $resourceGroupName = $_.ResourceGroupName        
+                $storageAccountName = $_.StorageAccountName               
+                try
+                {
+                    $storageAccount = Get-AzStorageAccount  -ResourceGroupName $resourceGroupName -Name $storageAccountName -ErrorAction SilentlyContinue
+                    $storageAccountResources += $storageAccount | Select-Object @{N='StorageAccountName';E={$_.StorageAccountName}},
+                                                                @{N='ResourceGroupName';E={$_.ResourceGroupName}},
+                                                                @{N='PrimaryLocation';E={$_.PrimaryLocation}},
+                                                                @{Name ='MinimalTlsVersion';Expression={ (Get-AzStorageAccount -Name $_.StorageAccountName -ResourceGroupName $_.ResourceGroupName).MinimumTlsVersion }}
```
Comment (aakashpahwa-msft) [inline Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1:355]: Same question as above.

File: Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1 [hunk 4]
```diff
@@ -0,0 +1,959 @@
...
+        Write-Host $([Constants]::DoubleDashLine)
+        break
+    }
+
+    Write-Host "Found $($totalstorageAccountResources) Storage Account(s)." -ForegroundColor $([Constants]::MessageType.Update)
+    Write-Host $([Constants]::SingleDashLine)
+ 
+     
+    # Includes Storage Accounts where minimal required TLS version is set  
+    $StorageAccountsWithReqMinTLSVersion = @()
```
Comment (aakashpahwa-msft) [inline Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1:381]: Why have we defined this? Is there any use?

File: Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1 [hunk 5]
```diff
@@ -0,0 +1,959 @@
...
+        $storageAccount = $_
+        $storageAccountName = $_.StorageAccountName
+        $resourceGroupName = $_.ResourceGroupName
+        $minimalTlsVersionBefore = $_.MinimalTlsVersionBefore
+        $minimalTlsVersionAfter = $_.MinimalTlsVersionAfter
+
+        try
+        {
+            $StorageAccount = ( Get-AzStorageAccount -StorageAccountName $storageAccountName  -ResourceGroupName $resourceGroupName -ErrorAction SilentlyContinue) 
+            $StorageAccountTls = (Get-AzStorageAccount -Name $storageAccountName -ResourceGroupName $resourceGroupName).MinimumTlsVersion
```
Comment (aakashpahwa-msft) [inline Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1:782]: Two Get calls again. Confirm once whether this is needed.

File: Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1 (hunk 1 above)
Comment (abhikash01) [reply Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1]: Replaced two calls with one. 
The parameter MinimalTlsVersion is coming when we explicitly get it using . operator. But not coming in as paramater when we run Get-AzStorageAccount

File: Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1 (hunk 2 above)
Comment (abhikash01) [reply Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1]: Replaced two calls with one. 
The parameter MinimalTlsVersion is coming when we explicitly get it using . operator. But not coming in as paramater when we run Get-AzStorageAccount

File: Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1 (hunk 3 above)
Comment (abhikash01) [reply Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1]: Replaced with one

File: Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1 (hunk 5 above)
Comment (abhikash01) [reply Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1]: Removed. Not required with new logic

File: Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1 (hunk 4 above)
Comment (abhikash01) [reply Scripts/RemediationScripts/Remediate-SetStorageAccountMinReqTLSVersion.ps1]: Not required actually. Removed

File: Scripts/RemediationScripts/ControlsEligibleForRemediationThroughUI.md [hunk 6]
```diff
@@ -29,6 +29,8 @@
 
 14. [Azure_SQLDatabase_DP_Use_Secure_TLS_Version_Trial](ControlsEligibleForRemediationThroughUI.md#14-Azure_SQLDatabase_DP_Use_Secure_TLS_Version_Trial)
 
+15. [Azure_Storage_DP_Use_Secure_TLS_Version_Trial](ControlsEligibleForRemediationThroughUI.md#14-Azure_Storage_DP_Use_Secure_TLS_Version_Trial)
```
Comment (aakashpahwa-msft) [inline Scripts/RemediationScripts/ControlsEligibleForRemediationThroughUI.md:32]: Change 14 --> 15 in "...ThroughUI.md#14-Azure_Sto..."
It's making the click feature on markdown not work.

File: Scripts/RemediationScripts/ControlsEligibleForRemediationThroughUI.md [hunk 7]
```diff
@@ -213,3 +215,16 @@ Use Approved TLS Version in SQL Server
...
+## 15. Azure_Storage_DP_Use_Secure_TLS_Version_Trial
+
+### Display Name
+Use Approved TLS Version in Storage Account
+
+### Link to Bulk Remediation Script (BRS)
+[Remediate-Azure_Storage_DP_Use_Secure_TLS_Version_Trial](Remediate-SetStorageAccountMinReqTLSVersion.ps1)
+
+### Minimum permissions required to run the script
+Owner or higher priviliged role on the Service Fabric(s)
```
Comment (aakashpahwa-msft) [inline Scripts/RemediationScripts/ControlsEligibleForRemediationThroughUI.md:228]: Service Fabric to Storage

Comment (abhikash01) [reply Scripts/RemediationScripts/ControlsEligibleForRemediationThroughUI.md:228]: Changed

File: Scripts/RemediationScripts/ControlsEligibleForRemediationThroughUI.md (hunk 6 above)
Comment (abhikash01) [reply Scripts/RemediationScripts/ControlsEligibleForRemediationThroughUI.md:32]: Changed

**Review by aakashpahwa-msft — APPROVED** [review]
LGTM

File: Scripts/RemediationScripts/ControlsEligibleForRemediationThroughUI.md [hunk 8]
```diff
@@ -213,3 +215,16 @@ Use Approved TLS Version in SQL Server
...
+## 15. Azure_Storage_DP_Use_Secure_TLS_Version_Trial
+
+### Display Name
+Use Approved TLS Version in Storage Account
+
+### Link to Bulk Remediation Script (BRS)
+[Remediate-Azure_Storage_DP_Use_Secure_TLS_Version_Trial](Remediate-SetStorageAccountMinReqTLSVersion.ps1)
+
+### Minimum permissions required to run the script
+Owner or higher priviliged role on the Storage Account(s)
```
Comment (Aboli-msft) [inline Scripts/RemediationScripts/ControlsEligibleForRemediationThroughUI.md:228]: Should it be contributor

**Review by Aboli-msft — APPROVED** [review]

**Pull request merged**